File: backend/app/api/feeds.py
from fastapi import APIRouter
from typing import List
from app.models.feeds import CivicFeed
from app.database.mongodb import get_database

router = APIRouter()

MOCK_FEEDS = [
    {
        "id": "feed-aws-04",
        "feedNumber": "AWS-04",
        "name": "Ashok Nagar Rain Gauge",
        "protocol": "MQTT / TELEMETRY",
        "status": "LIVE",
        "statusDetail": "Streaming normally (1 Hz cadence)",
        "frequency": "1 Hz",
        "latencyMs": 14,
        "jitterMs": 2,
        "packetsDropped": 0,
        "samplePayload": {"precipitation": 48.2, "unit": "mm/h", "sensor_id": "AWS-04", "status": "LIVE"},
        "lastIngestSha256": "f2a8b9c1d8e7f6a5b4c3d2e1f0a9b8c7d6e5f4a3b2c1d0e9f8a7b6c5d4e3f2a1",
        "description": "Primary atmospheric water sensor located at Ashok Nagar sub-station.",
        "degradationImpact": "Loss reduces precipitation confidence score by 35%."
    },
    {
        "id": "feed-loop-d12",
        "feedNumber": "LOOP D-12",
        "name": "Smart Mobility Loop D-12",
        "protocol": "REST / GEOJSON",
        "status": "LIVE",
        "statusDetail": "Streaming normally (5 sec cadence)",
        "frequency": "0.2 Hz",
        "latencyMs": 42,
        "jitterMs": 5,
        "packetsDropped": 0,
        "samplePayload": {"speed_km_h": 6.4, "occupancy_pct": 91.8, "sensor_id": "LOOP-D-12", "status": "LIVE"},
        "lastIngestSha256": "a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2",
        "description": "Induction loop speed and occupancy sensor at Panch Batti arterial intersection.",
        "degradationImpact": "Loss forces fallback to transit speed interpolation."
    },
    {
        "id": "feed-sampark-181",
        "feedNumber": "SAMPARK 181",
        "name": "Sampark 181 Dispatch Cell",
        "protocol": "XML INGEST / REST",
        "status": "LIVE",
        "statusDetail": "Batch sync nominal (1 min interval)",
        "frequency": "0.016 Hz",
        "latencyMs": 120,
        "jitterMs": 18,
        "packetsDropped": 0,
        "samplePayload": {"complaint_count": 38, "categories": ["waterlogging", "stalled_vehicle"], "status": "LIVE"},
        "lastIngestSha256": "9f8e7d6c5b4a3f2e1d0c9b8a7f6e5d4c3b2a1f0e9d8c7b6a5f4e3d2c1b0a9f8e",
        "description": "Rajasthan 181 citizen grievance portal dispatch queue for municipal dispatches.",
        "degradationImpact": "Loss disables citizen complaint density anomaly validation."
    },
    {
        "id": "feed-jcts-bus",
        "feedNumber": "JCTSL BUS API",
        "name": "JCTSL GTFS-RT Bus Telemetry",
        "protocol": "PROTOBUF / HTTP",
        "status": "DEGRADED",
        "statusDetail": "HTTP 504 Gateway Timeout (Circuit Breaker Active)",
        "frequency": "0.1 Hz",
        "latencyMs": 11840,
        "jitterMs": 3920,
        "packetsDropped": 12,
        "samplePayload": {"route_9a_speed_kmh": 4.2, "health": "DEGRADED (SYNTHETIC PROXY)", "error": "HTTP 504 Gateway Timeout"},
        "lastIngestSha256": "7c6b5a4f3e2d1c0b9a8f7e6d5c4b3a2f1e0d9c8b7a6f5e4d3c2b1a0f9e8d7c6b",
        "description": "Jaipur City Transport Services Limited real-time bus GTFS feed.",
        "degradationImpact": "Switched to synthetic proxy velocity penalty based on induction loops."
    }
]

MOCK_DEGRADATION = [
  {
    "phase": "PHASE 1",
    "title": "Circuit Breaker Socket Isolation",
    "status": "COMPLETED",
    "time": "13:38:12 IST",
    "detail": "Detected 3 consecutive HTTP 504 timeouts from JCTSL GTFS server. Socket isolated."
  },
  {
    "phase": "PHASE 2",
    "title": "Synthetic Proxy Weight Substitution",
    "status": "COMPLETED",
    "time": "13:38:15 IST",
    "detail": "Substituted live GTFS transit speed with induction loop traffic velocity (-2.5 km/h penalty)."
  },
  {
    "phase": "PHASE 3",
    "title": "Bayesian Confidence Score Adjustment",
    "status": "COMPLETED",
    "time": "13:38:18 IST",
    "detail": "Adjusted multi-stream concordance confidence weight for public transit stream from 0.90 to 0.82."
  },
  {
    "phase": "PHASE 4",
    "title": "Automated Sentinel Re-Probe Loop",
    "status": "ACTIVE NOW",
    "time": "14:00:00 IST",
    "detail": "Pinging JCTSL backup gateway endpoint every 30 seconds for automatic recovery."
  }
]

@router.get("/feeds", response_model=List[CivicFeed])
async def get_feeds():
    db = get_database()
    count = await db.feeds.count_documents({})
    if count == 0:
        await db.feeds.insert_many(MOCK_FEEDS)
        
    feeds_cursor = db.feeds.find({})
    feeds = await feeds_cursor.to_list(length=100)
    return [CivicFeed(**feed) for feed in feeds]

@router.get("/feeds/degradation")
async def get_degradation():
    db = get_database()
    if await db.degradation.count_documents({}) == 0:
        await db.degradation.insert_many(MOCK_DEGRADATION)
    return await db.degradation.find({}, {"_id": 0}).to_list(length=100)

File: src/data/mock/feeds.ts
import { CivicFeed } from '../../types';

export const CIVIC_FEEDS: CivicFeed[] = [
  {
    id: 'feed-01',
    feedNumber: 'FEED 01',
    name: 'Meteorology · IMD Weather Mesonet',
    protocol: 'REST / GeoJSON Polling (10s)',
    status: 'LIVE',
    statusDetail: '200 OK · Nominal Cadence',
    frequency: '10 sec',
    latencyMs: 142,
    jitterMs: 12,
    packetsDropped: 0,
    samplePayload: {
      station_id: "IMD-AWS-04",
      latitude: 26.9142,
      longitude: 75.8080,
      precipitation_rate_mm_hr: 48.2,
      accumulated_daily_mm: 72.4,
      relative_humidity: 94.2,
      barometric_pressure_hpa: 1004.2,
      timestamp_utc: "2026-09-24T08:30:00Z"
    },
    lastIngestSha256: '9a7f4c12d8e05b382901aaec9f41b9e28d7c4a1b8e23f048d21b7e61a49c30f1',
    description: 'High-density automated weather station (AWS) array across Jaipur Municipal Corporation boundaries.',
    degradationImpact: 'If offline, falls back to Doppler Radar extrapolation with -0.15 epistemic confidence penalty.'
  },
  {
    id: 'feed-02',
    feedNumber: 'FEED 02',
    name: 'Mobility · Smart Sensor Grid (Loop & Radar)',
    protocol: 'MQTT / Protocol Buffers (5s)',
    status: 'LIVE',
    statusDetail: '200 OK · Nominal Cadence',
    frequency: '5 sec',
    latencyMs: 88,
    jitterMs: 6,
    packetsDropped: 2,
    samplePayload: {
      loop_id: "LOOP-D-12",
      junction_name: "Panch Batti / M.I. Road",
      average_speed_kmh: 6.4,
      occupancy_pct: 91.8,
      vehicle_count_5m: 142,
      queue_length_meters: 320,
      timestamp_utc: "2026-09-24T08:30:05Z"
    },
    lastIngestSha256: '4f8812cbb0923e11a24d557f9208a0e371b83d1c479e0234a5d891b2c6499e12',
    description: 'In-pavement inductive loops and overhead radar sensors deployed by Jaipur Traffic Police & Smart City Ltd.',
    degradationImpact: 'If offline, falls back to Google Maps Distance Matrix API or historical time-of-day speed curves.'
  },
  {
    id: 'feed-03',
    feedNumber: 'FEED 03',
    name: 'Citizenry · Sampark 181 & 311 Municipal Portal',
    protocol: 'Webhook / Event-Driven WebSocket',
    status: 'LIVE',
    statusDetail: '200 OK · Streaming',
    frequency: 'Streaming',
    latencyMs: 210,
    jitterMs: 18,
    packetsDropped: 0,
    samplePayload: {
      ticket_id: "SMP-2024-81920",
      category: "CIVIC_WATERLOGGING_SEVERITY_3",
      reported_location: "Near Raj Mandir Cinema, M.I. Road",
      geo: [26.9152, 75.8078],
      caller_verified: true,
      sentiment_urgency: "HIGH",
      timestamp_utc: "2026-09-24T08:29:48Z"
    },
    lastIngestSha256: 'b3378901ae2b4c568912ef001847cbb592a3df41920a8c41238914b7e8894210',
    description: 'Centralized government grievance helpline (Rajasthan Sampark 181) and smart civic app dispatches.',
    degradationImpact: 'If offline, citizen signal is suspended; platform relies strictly on hardware telemetry.'
  },
  {
    id: 'feed-04',
    feedNumber: 'FEED 04',
    name: 'Public Transit · JCTSL Bus Fleet Telemetry',
    protocol: 'GTFS-Realtime (30s Polling)',
    status: 'OFFLINE',
    statusDetail: 'HTTP 504 Gateway Timeout · Protocol Alert',
    frequency: '30 sec (Stalled)',
    latencyMs: 12400,
    jitterMs: 4200,
    packetsDropped: 84,
    samplePayload: {
      route_id: "JCTSL-9A",
      bus_id: "RJ-14-PC-8921",
      last_known_speed: 4.2,
      last_gps_fix: "2026-09-24T08:08:12Z",
      status: "PACKET_TIMEOUT_ERROR_504",
      proxy_estimated_speed: 4.2
    },
    lastIngestSha256: 'c01198f123bc891042aa331ef89410bd2847a119854e89120bcdef4561239841',
    description: 'City bus GPS tracking network managed by Jaipur City Transport Services Limited.',
    degradationImpact: 'Currently in Phase 2 Inductive Loop Proxy Synthesis. Epistemic transparency penalty -0.08 applied.'
  }
];

export const DEGRADATION_PROTOCOL_STEPS = [
  {
    phase: 'PHASE 1',
    title: 'Circuit Breaker Socket Isolation',
    status: 'COMPLETED',
    time: '13:38:12 IST',
    detail: 'Detected 3 consecutive HTTP 504 timeouts from JCTSL GTFS server. Automated circuit breaker isolated feed to prevent blocking the ingestion pipeline.'
  },
  {
    phase: 'PHASE 2',
    title: 'Inductive Loop Proxy Synthesis',
    status: 'ACTIVE NOW',
    time: '13:38:15 IST',
    detail: 'Corridor speed is synthetically reconstructed using physical road induction loops (Loop D-12) & historical transit deceleration curves.'
  },
  {
    phase: 'PHASE 3',
    title: 'Epistemic Transparency Penalty',
    status: 'ENFORCED',
    time: '13:38:15 IST',
    detail: 'Overall model concordance penalised by -0.08. Uncertainty bounds widened on all transit-derived downstream inferences.'
  },
  {
    phase: 'PHASE 4',
    title: 'Civic Disclaimer Broadcast',
    status: 'BROADCASTING',
    time: '13:38:16 IST',
    detail: 'Public and municipal dashboards annotated with clear epistemic banner: "Transit metrics are synthetic estimates based on traffic loops."'
  }
];
